### PR TITLE
Refactor of Owner and Lessee ptrs [14325]

### PR DESCRIPTION
**Pull request by jparisu**
Signed-off-by: jparisu <javierparis@eprosima.com>

### CODE REVIEW

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

File: ddsrouter_utils/include/ddsrouter_utils/memory/owner_ptr.hpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+ * This object can access the data by operators -> and * and will throw an exception in case the data is not
+ * valid anymore (it has been deleted from the owner). While using the data, the owner will not be able to remove it.
+ * This object can also "lock" the data, creating a \c GuardedPtr that will force the data to not be removed
+ * while the object exist.
+ *
+ * - GUARDED PTR
+ * This object force the internal data to exist while this object exists (locks de data destruction).
+ * Thus, it is recommended to use this object as soon as possible and remove it afterwards, letting the ownership
+ * again to the owner of the data.
+ *
```
[the remark this answers is not in the document]
Comment (jparisu) [reply ddsrouter_utils/include/ddsrouter_utils/memory/owner_ptr.hpp:48]: `GuardedPtrData` is explained after LESSEE, line 45
`InternalPtrData` is not meant to be known neither used by the user, so the explanation must be in its own header

File: ddsrouter_utils/include/ddsrouter_utils/memory/impl/OwnerPtr.hpp [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+namespace eprosima {
+namespace ddsrouter {
+namespace utils {
+
+/**
+ * Class that contains a reference (ptr) for an object whose ownership is \c this .
+ * When \c this is destroyed, the object is destroyed too (using the deleter function given).
+ * In order to use the object from another context, it must be by using a \c LesseePtr ,
+ * that is allowed to use it and assures the object is not destroyed while it is being used.
+ * But LesseePtr does not own the object, which could only be destroyed when no lessee is using it.
```
[the remark this answers is not in the document]
Comment (jparisu) [reply ddsrouter_utils/include/ddsrouter_utils/memory/impl/OwnerPtr.hpp:41]: rephrased

File: ddsrouter_utils/include/ddsrouter_utils/exception/ValueAccessException.hpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+#define _DDSROUTERUTILS_EXCEPTIONS_VALUEACCESSEXCEPTION_HPP_
+
+#include <ddsrouter_utils/exception/Exception.hpp>
+
+namespace eprosima {
+namespace ddsrouter {
+namespace utils {
+
+/**
+ * @brief Exception thrown when not trying to reference a non valid value.
```
[the remark this answers is not in the document]
Comment (jparisu) [reply ddsrouter_utils/include/ddsrouter_utils/exception/ValueAccessException.hpp:29]: repeated

**Review by juanlofer-eprosima — APPROVED** [review]
LGTM

**Pull request merged**